src/sentry/static/sentry/app/utils/discover/genericDiscoverQuery.tsx

```diff
@@ -0,0 +1,207 @@
+  componentDidUpdate(prevProps: Props<T, P>) {
+    // Reload data if we aren't already loading,
+    const refetchCondition = !this.state.isLoading && this._shouldRefetchData(prevProps);
+
+    // or if we've moved from an invalid view state to a valid one,
+    const eventViewValidation =
+      prevProps.eventView.isValid() === false && this.props.eventView.isValid();
+
+    const shouldRefetchExternal = this.props.shouldRefetchData
+      ? this.props.shouldRefetchData(prevProps, this.props)
+      : false;
+
+    if (refetchCondition || eventViewValidation || shouldRefetchExternal) {
+      this.fetchData();
+    }
+  }
+
+  getPayload(props: Props<T, P>) {
+    if (this.props.getRequestPayload) {
+      return this.props.getRequestPayload(props);
+    }
+    return props.eventView.getEventsAPIPayload(props.location);
+  }
+
+  _shouldRefetchData = (prevProps: Props<T, P>): boolean => {
+    const thisAPIPayload = this.getPayload(this.props);
+    const otherAPIPayload = this.getPayload(prevProps);
+
+    return (
+      !isAPIPayloadSimilar(thisAPIPayload, otherAPIPayload) ||
+      prevProps.limit !== this.props.limit ||
+      prevProps.cursor !== this.props.cursor
+    );
+  };
+
+  fetchData = () => {
+    const {
+      api,
+      beforeFetch,
+      afterFetch,
+      eventView,
+      orgSlug,
+      route,
+      limit,
+      cursor,
+    } = this.props;
+
+    if (!eventView.isValid()) {
+      return;
+    }
+
+    const url = `/organizations/${orgSlug}/${route}/`;
+    const tableFetchID = Symbol(`tableFetchID`);
+    const apiPayload: Partial<EventQuery & LocationQuery> = this.getPayload(this.props);
+
+    this.setState({isLoading: true, tableFetchID});
+
+    if (limit) {
+      apiPayload.per_page = limit;
+    }
+    if (cursor) {
+      apiPayload.cursor = cursor;
+    }
+
+    beforeFetch?.(api);
+
+    api
+      .requestPromise(url, {
+        method: 'GET',
+        includeAllArgs: true,
+        query: {
+          // marking apiPayload as any so as to not cause typescript errors
+          ...(apiPayload as any),
+        },
+      })
+      .then(([data, _, jqXHR]) => {
+        if (this.state.tableFetchID !== tableFetchID) {
+          // invariant: a different request was initiated after this request
+          return;
+        }
+
+        const tableData = afterFetch ? afterFetch(data, this.props) : data;
+
+        this.setState(prevState => ({
+          isLoading: false,
+          tableFetchID: undefined,
+          error: null,
+          pageLinks: jqXHR ? jqXHR.getResponseHeader('Link') : prevState.pageLinks,
+          tableData,
+        }));
+      })
+      .catch(err => {
+        this.setState({
+          isLoading: false,
+          tableFetchID: undefined,
+          error: err?.responseJSON?.detail ?? null,
+          tableData: null,
+        });
+      });
+  };
+
+  render() {
+    const {isLoading, error, tableData, pageLinks} = this.state;
+
+    const childrenProps: ChildrenProps<T> = {
```

Comment:
```suggestion
    const childrenProps: GenericChildrenProps<T> = {
```
Should this be GenericChildrenProps?